File: src/jk_5/nailed/map/stats/StatManager.java
package jk_5.nailed.map.stats;

import com.google.common.collect.Lists;
import net.minecraft.src.TileEntityCommandBlock;

import java.util.Iterator;
import java.util.List;

/**
 * No description given
 *
 * @author jk-5
 */
public class StatManager {

    private final List<CommandBlock> commandBlocks = Lists.newArrayList();

    public void updateCommandBlock(TileEntityCommandBlock commandBlock) {
        if (stripFormatting(commandBlock.getCommandSenderName()).startsWith("?")) {
            for (CommandBlock block : this.commandBlocks) {
                if (block.getTileEntity() == commandBlock) return;
            }
            this.commandBlocks.add(new CommandBlock(commandBlock));
        } else {
            Iterator<CommandBlock> it = this.commandBlocks.iterator();
            while (it.hasNext()) {
                CommandBlock block = it.next();
                if (block.getTileEntity() == commandBlock) {
                    it.remove();
                    return;
                }
            }
        }
    }

    /**
     * WARNING: this actually sleeps a short time! Use with care!
     */
    public void triggerStat(String stat) throws InterruptedException {
        this.enableStat(stat);
        Thread.sleep(300);
        this.disableStat(stat);
    }

    public void enableStat(String stat) {
        for (CommandBlock block : this.commandBlocks) {
            if (block.getListenerName().equalsIgnoreCase("?" + stat)) {
                block.setRedstoneOutput(15);
            }
        }
    }

    public void disableStat(String stat) {
        for (CommandBlock block : this.commandBlocks) {
            if (block.getListenerName().equalsIgnoreCase("?" + stat)) {
                block.setRedstoneOutput(0);
            }
        }
    }

    static String stripFormatting(String input) {
        return input.replaceAll("§[0-9a-fA-F]+", "");
    }
}

File: src/main/scala/jk_5/nailed/map/stats/StatManager.scala
package jk_5.nailed.map.stats

import scala.collection.mutable.ArrayBuffer
import net.minecraft.tileentity.TileEntityCommandBlock

/**
 * No description given
 *
 * @author jk-5
 */
object StatManager {
  private final val commandBlocks = ArrayBuffer[CommandBlock]()

  def updateCommandBlock(block: TileEntityCommandBlock){
    if(this.stripFormatting(block.getCommandSenderName).startsWith("?")){
      if(!this.commandBlocks.exists(_.getTileEntity == block)){
        this.commandBlocks += new CommandBlock(block)
      }
    }else this.commandBlocks.filter(_.getTileEntity == block).foreach(b => this.commandBlocks -= b)
  }

  /**
   * WARNING: this actually sleeps a short time! Use with care!
   */
  def triggerStat(stat: String){
    this.enableStat(stat)
    Thread.sleep(300)
    this.disableStat(stat)
  }
  def enableStat(stat: String) = this.commandBlocks.filter(_.getListenerName.equalsIgnoreCase("?" + stat)).foreach(_.setRedstoneOutput(15))
  def disableStat(stat: String) = this.commandBlocks.filter(_.getListenerName.equalsIgnoreCase("?" + stat)).foreach(_.setRedstoneOutput(0))
  @inline def stripFormatting(input: String) = input.replaceAll("§[0-9a-fA-F]+", "")
}
